paddle/fluid/eager/backward.cc

```diff
@@ -74,14 +82,189 @@ std::unordered_map<GradNodeBase*, int> getInDegreeMap(
+void UpdateGraphInfo(
+    std::unordered_map<GradNodeBase*, AutogradMeta*>*
+        target_nodes_inputmeta_map,
+    std::unordered_map<GradNodeBase*, GradNodeBase*>* depending_nodes,
+    std::unordered_set<GradNodeBase*>* potential_stop_nodes) {
+  // according depending_nodes to updated potential_sotp_nodes,
+  // make sure the path from root to target_node is ok
+  VLOG(6) << "Update GraphInfo";
+  std::queue<GradNodeBase*> queue;
+  for (auto& target_nodes_inputmeta_pair : *target_nodes_inputmeta_map) {
+    queue.emplace(target_nodes_inputmeta_pair.first);
+  }
+
+  while (!queue.empty()) {
+    auto* target_node = queue.front();
+    queue.pop();
+    if ((*depending_nodes)[target_node]) {
+      auto* precedding_node = (*depending_nodes)[target_node];
+      queue.emplace(precedding_node);
+      if (potential_stop_nodes->find(precedding_node) !=
+          potential_stop_nodes->end()) {
+        potential_stop_nodes->erase(precedding_node);
+      }
+    }
+  }
+}
+
+// Get Graph Info Betweent input target gradnode and outputs
+// Record the depending_nodes and potential_stop_nodes
+void GetGraphInfoBetweenTargets(
+    const std::queue<GradNodeBase*>& init_queue,
+    std::unordered_map<GradNodeBase*, AutogradMeta*>*
+        target_nodes_inputmeta_map,
+    std::unordered_map</*child node*/ GradNodeBase*,
+                       /*father node*/ GradNodeBase*>* depending_nodes,
+    std::unordered_set<GradNodeBase*>* potential_stop_nodes) {
+  VLOG(6) << "Run In GetGraphInfoBetweenTargets";
+  // Calculate in_degree for each node
+  std::unordered_map<GradNodeBase*, int> node_in_degree_map;
+
+  // Copy nodes
+  std::queue<GradNodeBase*> queue = init_queue;
+  std::unordered_set<GradNodeBase*> visited;
+
+  // Visit each node exactly once in any order
+  while (!queue.empty()) {
+    GradNodeBase* node = queue.front();
+    queue.pop();
+
+    if (visited.count(node)) {
+      continue;
+    }
+    visited.insert(node);
+
+    // Check node is target_nodes or not, if node is not target_node,
+    // all the next_node will be marked in potential_stop_nodes
+    bool is_potential_stop_nodes = false;
+    if (target_nodes_inputmeta_map->find(node) !=
+        target_nodes_inputmeta_map->end()) {
+      is_potential_stop_nodes = true;
+    }
+
+    // Find and append next nodes
+    const std::vector<std::vector<Edge>>& edges = node->GetEdges();
+    for (const auto& edge_list : edges) {
+      for (const Edge& edge : edge_list) {
+        GradNodeBase* next_node = edge.GetMutableGradNode().get();
+
+        // Next node could be nullptr if it is leaf tensor with no
+        // AccumulationNode attached
+        // Or it could also originated from dispensable inputs
+        if (!next_node) continue;
+
+        // if node not in input_target_nodes,
+        // all the next_nodes of current node will be inserted to
+        // potential_stop_node
+        if (is_potential_stop_nodes) {
+          potential_stop_nodes->emplace(next_node);
```

Comment:
Unrecommended use of "emplace", refer to the other comment on "emplace_back and std::move"

Reply:
Thx, but std::unordered_set<GradNodeBase*>* potential_stop_nodes could not call emplace_back.